fix: OUMVLP-Pose pkl data temporal discontinuity

Ahzyuan opened the pull request:
Hi 👋,

Thanks for your wonderful work, which has truly advanced the field of gait recognition.

## 🐞 Bug statement

However, I found a bug during the use. That is, in `OUMVLP-Pose` pkl data extraction(code shown below), the origin json file may store several skeletons but not one. And in the script, the `Line 151` will choose the first skeleton to be the pose data for this frame by default.

https://github.com/ShiqiYu/OpenGait/blob/c42f2f8c0724141e37be97614843a7347bd4ae61/datasets/pretreatment.py#L144-L154

However, this behavior might be wrong and cause the **spatial discontinuity** in the final pose sequence which may definitely **harm the downstream algorithm accuracy and performance**! Here is an example:

<details><summary>Spatial Discontinuity Example (Click Me)</summary>
<p>

- Sequence: `OUMVLP-Pose/Alphapose/07052/01/180`
- [Sequence visualization download link](https://drive.google.com/file/d/1VkyFlxhCwJoOPSSmt10YJ1W0zPfFbDvK/view?usp=drive_link)

If the existing script is used to extract the pose pkl data, and draw it on the corresponding silhouette frame, you will see the spatial discontinuity phenomenon like this:

![image](https://github.com/user-attachments/assets/690c13be-0986-4d92-89fe-3132f939e8e3)

You can see that the skeletons have a **spatial translation** and are **not attached** to the human body in hardly **half of the frames**!

</p>
</details> 

## 🔨 Fix statement

I have fixed this bug by calculating the **skeleton-silu matching score** which uses the sum of the silu pixel intensity on all joint coordinates. That is the [`pose_silu_match_score()` function](https://github.com/ShiqiYu/OpenGait/pull/280/files#diff-4df89e28d411aef30a3e77ff46422083a74c154cf5473b94f9bac37b20112337R141-R153) in the change.

**I have tested the code and successfully extracted a new pose pkl dataset**. Using these pkl files, I redrew the visualization of the above sequence, and the effect is remarkable:

<details><summary>Re-visualization using new pkl data extracting by changed script (Click Me)</summary>
<p>

- Sequence: `OUMVLP-Pose/Alphapose/07052/01/180`
- [Sequence visualization download link](https://drive.google.com/file/d/186Yp-p0TgzSKPIRopSQ9gtOWOKlB7b9q/view?usp=drive_link)

![image](https://github.com/user-attachments/assets/5c0b108d-ecab-449b-bb7e-d0fa86e55579)

</p>
</details> 

## 🧩 Others

I have updated the corresponding command for extracting `OUMVLP-Pose` pkl data in `datasets/OUMVLP/README.md`.

jdyjjj commented:
> Awesome results. The fix looks good to me. @ChaoFan996 @jdyjjj @broweigit @HappyDeepLearning any comment?
Dealing with it this way effectively fixes cases we hadn’t considered before. I think it’s really good.

Ahzyuan commented:
@chuanfushen @broweigit @jdyjjj, thank you for your recognition 🙏!

I understand your request. Currently, my logic not only requires **additional calculations** but also **extra image reading IO** (including disk and memory, which may cause serious bottlenecks). Most importantly, users need to **apply for both silhouette and skeleton datasets simultaneously**. Since the application process takes a long time, this will directly affect the user experience.

I've thought about your suggestion 🤔, and here are **two** methods I can come up with for now:

<details><summary>Method 1: Save and reuse matching result</summary>
<p>

1. **Pipeline**
    - Based on my current modifications → 
    - Generate the selection index for each frame of each sequence →
    - Make it publicly available for free. **Users can directly load the index after downloading it**. 

2. **Pros and Cons**
    - Pros:
        1. For each sequence, users only need to read an index array and **do not need to read images**, thereby reducing disk IO and memory consumption.
        2. Users **do not need to apply for the silhouette dataset**
        3. The effect remains **as accurate as the approach now**
    - Cons: still need to load extra reference data
 
</p>
</details> 

<details><summary>Method 2: Inter-frame multi-target tracking</summary>
<p>

Inspired by the heuristic algorithm you mentioned, imagine such a **generalized and universal** scenario: 

> in a certain video, in addition to the person being filmed, the crew members at the edges occasionally enter or exit the frame. At this time, the JSON data of each frame will contain a random number of skeletons, and they will be out of order. 

In this case, we need to **combine these out-of-order data into multiple potential skeleton sequences** and select the one with the **largest movement range** to be the final sequence of this video. This is actually a typical case in the field of <ins>target tracking</ins>, so we can use the classic `Kalman filter` and `Hungarian algorithm` to accomplish this task:

1. **Mechanism**
    - Input: `JSON` poses data for each frame
    - Output: The pose sequence with maximum movement distance

2. **Pipeline**
    1. For each skeleton detected in the first frame, initialize a `Kalman filter` (using the midpoint of the spine as the initial state).
    2. for subsequent frames:
        - For all tracked skeletons, use their own `Kalman filter` to predict the position of their center points in the current frame.
        - Calculate the Euclidean distances between the predicted center point and each skeleton center point in the current frame to obtain a distance matrix.
        - With the distance matrix, the `Hungarian algorithm` is used to match the skeleton of the previous frame with that of the current frame:
            1. For successfully matched targets, update its `Kalman filter` using the skeleton center points matched in the current frame.
            2. Unmatched detection results shall be regarded as new targets, and a new `Kalman filter` shall be created.
            3. For tracked targets that have not been matched for a preset long time, delete them

    3. After all frames have been processed, calculate the **linear movement distance** of the center point for each trajectory, and select the trajectory with the **maximum movement distance** to output as the picked sequence of this video.

3. **Pseudo Code** 

    <details><summary>Click to expand (BTW, you can preview the python implement by giving it to AI)</summary>
    <p>
    
    ```pseude-code
    FUNCTION compute_skeleton_center(skeleton) RETURNS center_point // Calculate the central axis of the spinal skeleton
    FUNCTION calculate_euclidean_distance(p1, p2) RETURNS distance // Calculate the Euclidean distance between two points
    FUNCTION initialize_kalman_filter(initial_position) RETURNS kalman_filter // Initialize the Kalman filter
    FUNCTION kalman_predict(filter) RETURNS predicted_state // Kalman Filter Prediction
    FUNCTION kalman_update(filter, measurement) RETURNS updated_filter // Kalman Filter Update
    FUNCTION hungarian_match(cost_matrix) RETURNS matches, unassigned_poses, unassigned_tracks // Hungarian algorithm, returning skeleton-trajectory matching pairs, unmatched     skeletons, and unmatched trajectories
    FUNCTION sort(array, reverse=False) RETURNS sorted_array // Sort an array in ascending order or descending order(reverse=True)
    
    FUNCTION main(frames_jsons):
        DECLARE MAX_MISSED_FRAMES AS INTEGER ← 5  // Maximum Allowable Mismatched Frames
        DECLARE tracked_skeletons AS ARRAY  // Store the Kalman filters and trajectories for all tracked skeletons
        
        // Initialize Kalman filters for all skeletons in the first frame
        FOR pose_idx, pose IN ENUMERATE(frames_jsons[0]["people"]) DO
            center ← compute_skeleton_center(pose["pose_keypoints_2d"])
            kf ← initialize_kalman_filter(center)
            APPEND TO tracked_skeletons {
                "kalman_filter": kf,
                "trajectory": [center],
                "inframe_pose_idx": [pose_idx],
                "last_seen": 0  // Record the last matching frame index
            }
        END FOR
        
        // Process subsequent frames
        FOR frame_idx FROM 1 TO LENGTH(frames_jsons)-1 DO
            DECLARE predicted_centers AS ARRAY // Store the predicted center of the tracked skeletons (in the order of recording)
            DECLARE current_centers AS ARRAY // Store the center point of the skeleton in the current frame (in the order as in the JSON)
            DECLARE cost_matrix AS ARRAY // Distance Matrix (Rows: predicted center points of tracked skeletons, Columns: center points of skeletons in current frame)
            
            current_frame ← frames_jsons[frame_idx]
    
            // Predict the center points of all tracked skeletons in current time step
            FOR i FROM 0 TO LENGTH(tracked_skeletons)-1 DO
                track ← tracked_skeletons[i]
                pred_center ← kalman_predict(track["kalman_filter"])
                APPEND pred_center TO predicted_centers
            END FOR
            
            // Calculate the center points of all skeletons present in current frame
            FOR EACH pose IN current_frame["people"] DO
                center ← compute_skeleton_center(pose["pose_keypoints_2d"])
                APPEND center TO current_centers
            END FOR
            
            // Calculate the distance matrix
            FOR EACH pred IN predicted_centers DO
                DECLARE row AS ARRAY
                FOR EACH det IN current_centers DO
                    distance ← calculate_euclidean_distance(pred, det)
                    APPEND distance TO row
                END FOR
                APPEND row TO cost_matrix
            END FOR
            
            // Use Hungarian algorithm to match the skeleton of the previous frames with that of the current frame
            matches, unassigned_poses, unassigned_tracks ← hungarian_match(cost_matrix)
            
            // Update matched tracks
            FOR track_idx, current_pose_idx IN ENUMERATE(matches) DO
                track ← tracked_skeletons[track_idx]
                pose_center ← current_centers[current_pose_idx]
    
                updated_kf ← kalman_update(track["kalman_filter"], pose_center)
                track["kalman_filter"] ← updated_kf
                
                APPEND pose_center TO track["trajectory"]
                APPEND current_pose_idx TO track["inframe_pose_idx"]
                track["last_seen"] ← frame_idx
            END FOR
            
            // Process the unmatched skeletons, they will be added to the tracking list as new targets.
            FOR EACH current_pose_idx IN unassigned_poses DO
                new_pose_center ← current_centers[current_pose_idx]
                new_kf ← initialize_kalman_filter(new_pose_center)
                APPEND TO tracked_skeletons {
                    "kalman_filter": new_kf,
                    "trajectory": [new_pose_center],
                    "inframe_pose_idx": [current_pose_idx],
                    "last_seen": frame_idx
                }
            END FOR
            
            // Process the unmatched trajectories, which indicate that the skeleton may disappear in next 5 frames
            deprecated_tracks ← []
            FOR EACH track_idx IN unassigned_tracks DO
                IF frame_idx - tracked_skeletons[track_idx]["last_seen"] > MAX_MISSED_FRAMES THEN
                    APPEND track_idx TO deprecated_tracks // Mark the trajectories that have not been matched for more than 5 frames as deprecated
                END IF
            END FOR
    
    
            // Delete the deprecated trajectories
            FOR EACH track_idx IN sort(deprecated_tracks, reverse=True) DO
                DELETE tracked_skeletons[track_idx]
            END FOR
        END FOR
        
        // Calculate the linear movement distance for each skeleton trajectory
        max_distance ← 0
        pick_idx ← NULL
        
        FOR EACH track IN tracked_skeletons DO
            linear_distance ← calculate_euclidean_distance(
                track["trajectory"][0], 
                track["trajectory"][-1]
            )
            
            IF linear_distance > max_distance THEN
                max_distance ← linear_distance
                pick_idx ← track["inframe_pose_idx"]
            END IF
        END FOR
        
        // Select the trajectory with the maximum movement distance to output as the picked sequence
        pose_seq ← []
        FOR frame_idx FROM 0 TO LENGTH(frames_jsons)-1 DO
            pose_list ← frames_jsons[frame_idx]["people"]
            target_pose_idx ← pick_idx[frame_idx]
            APPEND pose_list[target_pose_idx]["pose_keypoints_2d"] TO pose_seq
        END FOR
    
        RETURN pose_seq
    END FUNCTION
    ```
    
    </p>
    </details> 

4. **Pros and Cons**
    - Pros: Completely no need for any extra reference info
    - Cons: 
        1. The code refactoring involves a large amount of work and its **implementation is more complex**
        2. The **computational load is greater** than that of the existing changes
        3. Due to the absence of a feature extraction mechanism, it is **impossible to establish the correlation between old and new trajectories**. If the frame sequence is discontinuous at this point, the subject being filmed will be viewed as different targets. **This will render part of the data in the old trajectory unavailable and significantly reduce the length of the final output sequence**.
        4. The hyper-parameters for the `Kalman filter` algorithm is simple and **mostly determined based on experience**, including the maximum mismatch length, process noise covariance (Q), measurement noise covariance (R), and so on.
        5. `Hungarian algorithm` matching is performed based on distance. **When two skeletons are close to each other, the tracking algorithm is prone to confusion.**

</p>
</details> 

What do you think? Waiting for your reply.

Regards,
Ahzyuan

Ahzyuan commented:
Hi, I have refactored my changes according to Method 1, and the final changes compared with the base branch are as follows.

## 🛠️ Change Statement

1. `datasets/pretreatment.py`: Compared to the base branch, **only the index loading logic has been added**
    - For each sequence, try to read its corresponding index file from the index library. If the file does not exist (e.g., accidentally deleted by the user), the pose selection behavior shall **revert to using the first detected one** as the default.
    - If the file is read successfully, a dictionary can be obtained. The key is the index of the frame containing multiple poses, and the value is the index of the pose to be selected. With this dictionary, when iterating through the `JSON` data of each frame, **a one-to-one mapping between each frame and the corresponding pose index it should select can be obtained**.

2. `datasets/OUMVLP/pose_index_extractor.py`：
    - This file is newly created to **extract the indices of the skeletons that should be selected for each frame in the sequence**, and the mechanism of calculating the sum of the pixel intensities of the joint points in the above context is the core of this file. 
    - Here, in order to **reduce the size of the final index file**, for each sequence, **I only store the index values of the frames where a selection occurs**. If the number of poses contained in a `JSON` file is ≤ 1, there is no need to do a selection, and thus its data does not need to be stored.

4. `datasets/OUMVLP/README.md`：Before the command for extracting the `OUMVLP-Pose` pkl file, I have added **two methods** to obtain the index file, and users can choose to either **download it directly** or **extract it themselves**.

## 🧪 Code Test

For this commit, I have verified the correctness of the code. I have selected another sequence with misalignment, and here is a comparison of the three methods up to now.

seq: `OUMVLP-Pose/Alphapose/07045/00/180`
<details>
<summary>index file content</summary>

```
{
    '0027': 0,
    '0028': 0,
    '0029': 0,
    '0030': 0,
    '0031': 0,
    '0032': 0,
    '0033': 0,
    '0034': 1,
    '0035': 1,
    '0036': 1,
    '0037': 1,
    '0038': 1,
    '0039': 1,
    '0040': 1,
    '0041': 1,
    '0042': 1,
    '0043': 1,
    '0044': 1,
    '0045': 1,
    '0046': 1,
    '0047': 1,
    '0048': 1,
    '0049': 1,
    '0050': 1,
    '0051': 1,
    '0052': 1,
    '0053': 1,
    '0054': 1
}
```

</details>

| original | real-time matching | load index |
| :-------:| :----------------: |:----------:|
| <img alt="origin" src="https://github.com/user-attachments/assets/c6723cdf-9b36-4648-8585-bd39c8308c05" /> | <img alt="real-time matching" src="https://github.com/user-attachments/assets/2de5ca12-f200-4d76-b95d-fe06d96b2bb8" /> | <img alt="load index" src="https://github.com/user-attachments/assets/341dbf5b-9986-447a-baa5-80a42835af86" /> |

It can be seemed that the result of `loading index` method is the **same** as that of the `real-time matching`, which proves the correctness of the code. In addition, I have also tested the modified `datasets/pretreatment.py`, and **successfully extracted the pkl dataset with the index files**.

---

**At this point, users can extract an accurate, no temporal misalignment `OUMVLP-Pose` pkl dataset with no extra calculation burden.** 

**Hoping this can make** `OpenGait` **a better framework** ✨.

Regards,
Ahzyuan

broweigit commented:
Great work!  @ChaoFan996 @jdyjjj @chuanfushen do you have further suggestions? If not, I think we could merge this commit.
